tools/state-viewer/src/replay_headers.rs

```diff
@@ -0,0 +1,156 @@
+    let epoch_manager = EpochManager::new_arc_handle(store, &near_config.genesis.config);
+
+    let replay_store = create_replay_store(home_dir, &near_config);
+    let epoch_manager_replay =
+        EpochManager::new_arc_handle(replay_store, &near_config.genesis.config);
+
+    for height in start_height..=end_height {
+        if let Ok(block_hash) = chain_store.get_block_hash_by_height(height) {
+            let header = chain_store.get_block_header(&block_hash).unwrap().clone();
+            tracing::trace!("Height: {}, header: {:#?}", height, header);
+
+            let header_info = get_block_header_info(&header, &chain_store, epoch_manager.as_ref())
+                .unwrap_or_else(|e| {
+                    panic!(
+                        "Failed to add chunk endorsements for block height {}: {:#}",
+                        header.height(),
+                        e
+                    )
+                });
+            epoch_manager_replay.add_validator_proposals(header_info).unwrap().commit().unwrap();
+
+            if epoch_manager
+                .is_last_block_in_finished_epoch(&block_hash)
+                .expect("Could not determine if block is last block in epoch")
+            {
+                let identifier = ValidatorInfoIdentifier::BlockHash(block_hash);
+                let validator_info = epoch_manager.get_validator_info(identifier.clone()).unwrap();
+                let validator_info_replay =
+                    epoch_manager_replay.get_validator_info(identifier).unwrap();
+
+                assert_eq!(validator_info.epoch_height, validator_info_replay.epoch_height);
+                tracing::info!(
+                    "Comparing validator infos for epoch height {} block height {}",
+                    validator_info.epoch_height,
+                    height
+                );
+
+                assert_eq!(validator_info, validator_info_replay);
+            }
+        }
+    }
+}
+
+/// Returns the [`BlockHeaderInfo`] corresponding to the header.
+/// This function may override the resulting [`BlockHeaderInfo`] based on certain protocol versions.
+fn get_block_header_info(
+    header: &BlockHeader,
+    chain_store: &ChainStore,
+    epoch_manager: &EpochManagerHandle,
+) -> Result<BlockHeaderInfo, Error> {
+    let mut header_info = BlockHeaderInfo::new(
+        &header,
+        chain_store.get_block_height(header.last_final_block()).unwrap(),
+    );
+
+    // Note(#11900): If chunk endorsements in block header is enabled but we have not yet populated the endorsements bitmap
+    // in the header yet, we generate chunk endorsement bitmap from the chunk endorsement signatures in the block body.
+    // TODO(#11900): Remove this code after ChunkEndorsementsInBlockHeader is stabilized.
+    let protocol_version = epoch_manager.get_epoch_protocol_version(header.epoch_id())?;
+    if ProtocolFeature::ChunkEndorsementsInBlockHeader.enabled(protocol_version)
+        && header.chunk_endorsements().is_none()
```

Comment:
I couldn't understand the condition. From the purpose set in PR, shouldn't we set the bitmap if this protocol version is **not** enabled?

Not critical for merging the PR, anyway.

Reply:
The part that sets the endorsement bitmap in the block header is not yet implemented. Thus `header.chunk_endorsements().is_none()` will be true until we implement and release it. This condition was to prevent overwriting the bitmap already in the header when that happens. But thinking twice, overwriting it is fine (it should set it to the same value). So I will change the condition to avoid the confusion.